x-pack/plugins/fleet/server/services/download_source.ts

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import type { SavedObjectsClientContract, SavedObject } from '@kbn/core/server';
+
+import {
+  DOWNLOAD_SOURCE_SAVED_OBJECT_TYPE,
+  DEFAULT_DOWNLOAD_SOURCE,
+  DEFAULT_DOWNLOAD_SOURCE_ID,
+} from '../constants';
+
+import type { DownloadSource, DownloadSourceAttributes, DownloadSourceBase } from '../types';
+import { DownloadSourceError } from '../errors';
+import { SO_SEARCH_LIMIT } from '../../common';
+
+import { agentPolicyService } from './agent_policy';
+import { appContextService } from './app_context';
+
+function savedObjectToDownloadSource(so: SavedObject<DownloadSourceAttributes>) {
+  const { source_id: sourceId, ...attributes } = so.attributes;
+
+  return {
+    id: sourceId ?? so.id,
+    ...attributes,
+  };
+}
+
+class DownloadSourceService {
+  public async get(soClient: SavedObjectsClientContract, id: string): Promise<DownloadSource> {
+    const soResponse = await soClient.get<DownloadSourceAttributes>(
+      DOWNLOAD_SOURCE_SAVED_OBJECT_TYPE,
+      id
+    );
+
+    if (soResponse.error) {
+      throw new Error(soResponse.error.message);
+    }
+
+    return savedObjectToDownloadSource(soResponse);
+  }
+
+  public async list(soClient: SavedObjectsClientContract) {
+    const downloadSources = await soClient.find<DownloadSourceAttributes>({
+      type: DOWNLOAD_SOURCE_SAVED_OBJECT_TYPE,
+      page: 1,
+      perPage: SO_SEARCH_LIMIT,
+      sortField: 'is_default',
+      sortOrder: 'desc',
+    });
+
+    return {
+      items: downloadSources.saved_objects.map<DownloadSource>(savedObjectToDownloadSource),
+      total: downloadSources.total,
+      page: downloadSources.page,
+      perPage: downloadSources.per_page,
+    };
+  }
+
+  public async create(
+    soClient: SavedObjectsClientContract,
+    downloadSource: DownloadSourceBase,
+    options?: { id?: string }
+  ): Promise<DownloadSource> {
+    const data: DownloadSourceAttributes = downloadSource;
+
+    // default should be only one
+    if (data.is_default) {
+      const defaultDownloadSourceId = await this.getDefaultDownloadSourceId(soClient);
+
+      if (defaultDownloadSourceId) {
+        await this.update(soClient, defaultDownloadSourceId, { is_default: false });
+      }
+    }
+    if (options?.id) {
+      data.source_id = options?.id;
+    }
+
+    const newSo = await soClient.create<DownloadSourceAttributes>(
+      DOWNLOAD_SOURCE_SAVED_OBJECT_TYPE,
+      data,
+      {
+        id: options?.id,
```

Comment:
Why do we have `source_id` and `id` as separate attributes, but we mirror their values here?

Reply:
I think I found the reason why in some saved objects it was used an additional "id" field: it's basically to have the option to set the "default" value  to an arbitrary value, in the case of this API I set it to `'fleet-default-download-source', otherwise it will be different every time fleet starts, since the id of a saved object cannot be set from outside. Having an additional `source_id` allows to circumvent this rule.